### PR TITLE
Joint Graph Runner 

### PR DESCRIPTION
**Pull request by SherlockNoMad**

### CODE REVIEW

File: torchtitan/experiments/joint_graph_runner/llama3/parallelize.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        dt_args = tuple(
+            DTensor.from_local(arg, self.parallel_dims.world_mesh["tp"], [Replicate()])
+            for arg in args
+        )
+
+        # RuntimeError('Sharding propagation failed for Op(op=aten.embedding.default, args_schema=Spec(S(0) on (2048, 256)), Spec((Shard(dim=0), Replicate()) on (16, 2048)) @ mesh: (2, 4))')
+        # dt_args = tuple(DTensor.from_local(arg, self.parallel_dims.world_mesh, [Shard(0), Replicate()]) for arg in args)
+
+        # RuntimeError('Sharding propagation failed for Op(op=aten.embedding.default, args_schema=Spec(S(0) on (2048, 256)), Spec(S(0) on (16, 2048)) @ mesh: (2,))')
+        # dt_args = tuple(DTensor.from_local(arg, self.parallel_dims.world_mesh["dp_shard"], [Shard(0)]) for arg in args)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/joint_graph_runner/llama3/parallelize.py]: here

**Comment by ezyang** [comment]
cc @bobrenjc93 @aorenste this might be a good way to look at compile on one rank, perhaps??

**Comment by SherlockNoMad** [comment]
> cc @bobrenjc93 @aorenste this might be a good way to look at compile on one rank, perhaps??

Yes, I have examples on how graphs are different on different ranks. 
We can investigate on how to parameterize them. 

P1975157784: rank0_autograd_function_0fea2786.py
P1975158481: rank1_autograd_function_28587623.py

File: torchtitan/models/attention.py [hunk 2]
```diff
@@ -85,8 +85,9 @@ def forward(
         v: torch.Tensor,
         scale: float | None = None,
     ) -> torch.Tensor:
-        block_mask = FlexAttention.block_masks[self.mask_key]
-        return FlexAttention.flex_attn(q, k, v, block_mask=block_mask, scale=scale)
+        with torch.fx.traceback.annotate({"compile_with_inductor": "flex_attention"}):
+            block_mask = FlexAttention.block_masks[self.mask_key]
+            return FlexAttention.flex_attn(q, k, v, block_mask=block_mask, scale=scale)
```
Comment (fegin) [inline torchtitan/models/attention.py]: The latest attention.py changes the argument. Block mask is now an argument of `forward()`.

Comment (fegin) [reply torchtitan/models/attention.py]: Is this the only part you need to annotate?

Comment (SherlockNoMad) [reply torchtitan/models/attention.py]: rebased and fixed. 

Yes, we only need to annotate FlexAttention for regional compilation.

File: torchtitan/models/attention.py [hunk 3]
```diff
@@ -67,14 +67,15 @@ def forward(
...
 
-        return FlexAttentionWrapper._compiled_flex_attn(
-            q,
-            k,
-            v,
-            block_mask=block_mask,
-            scale=scale,
-            return_lse=return_lse,
-        )
+        with torch.fx.traceback.annotate({"compile_with_inductor": "flex_attention"}):
```
Comment (SherlockNoMad) [inline torchtitan/models/attention.py]: remove.

File: torchtitan/experiments/compiler_toolkit/README.md [hunk 4]
```diff
@@ -21,3 +21,12 @@ NGPU=4 CONFIG_FILE=./torchtitan/models/deepseek_v3/train_configs/debug_model.tom
 ```shell
 NGPU=4 CONFIG_FILE=./torchtitan/models/deepseek_v3/train_configs/debug_model.toml ./run_train.sh --model.name compiler_toolkit.deepseek_v3 --compile.enable --parallelism.data_parallel_shard_degree=2 --parallelism.tensor_parallel_degree=2 --parallelism.expert_parallel_degree=2 --activation_checkpoint.mode none --model.flavor=debugmodel_flex_attn
 ```
+
+
+llama3
+
+**SimpleFSDP + TP**
+NGPU=8 CONFIG_FILE="./torchtitan/models/llama3/train_configs/debug_model.toml" with-proxy ./run_train.sh --model.name compiler_toolkit.llama3 --compile.enable --parallelism.data_parallel_shard_degree=2 --parallelism.tensor_parallel_degree=4
```
Comment (ruisizhang123) [inline torchtitan/experiments/compiler_toolkit/README.md]: torchtitan by default uses inductor backend. I think we need to add another arg `--compile.backend "aot_eager"` to fully override inductor backend; otherwise the loss will still be compiled by inductor [here](https://github.com/pytorch/torchtitan/blob/f51ab9fb161acbb801352105392df4eefa46bd0d/torchtitan/components/loss.py#L26-L32)

Comment (SherlockNoMad) [reply torchtitan/experiments/compiler_toolkit/README.md]: good catch

File: torchtitan/experiments/compiler_toolkit/README.md [hunk 5]
```diff
@@ -1,23 +1,35 @@
...
+  - fsdp bucketing/prefetching for comm/compute overlap
 - Run using the aot_compile_joint_with_descriptors API
 
-**SimpleFSDP + TP + EP**
+## DeepSeek v3
 
+**SimpleFSDP + TP + EP**
 ```shell
-NGPU=4 CONFIG_FILE=./torchtitan/models/deepseek_v3/train_configs/debug_model.toml ./run_train.sh --model.name compiler_toolkit.deepseek_v3 --compile.enable --parallelism.data_parallel_shard_degree=2 --parallelism.tensor_parallel_degree=2 --parallelism.expert_parallel_degree=2 --activation_checkpoint.mode none
+NGPU=4 CONFIG_FILE=./torchtitan/models/deepseek_v3/train_configs/debug_model.toml ./run_train.sh --model.name compiler_toolkit.deepseek_v3 --compile.enable --compile.backend "aot_eager" --parallelism.data_parallel_shard_degree=2 --parallelism.tensor_parallel_degree=2 --parallelism.expert_parallel_degree=2 --activation_checkpoint.mode none
```
Comment (yiming0416) [inline torchtitan/experiments/compiler_toolkit/README.md]: @SherlockNoMad I think we should remove `--compile.enable` as well as `--compile.backend` here.
Since we are applying the customized `CompiledModule`, we don't really want `torch.compile()` the model.

File: torchtitan/experiments/compiler_toolkit/llama3/parallelize.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+def parallelize_llama(
+    model: torch.nn.Module,
+    parallel_dims: ParallelDims,
+    job_config: JobConfig,
+) -> CompiledModule:
+
+    annotate_model()
+
+    # Disable torch.compile over the model in the compiler toolkit style workflow
+    with disable_compile(job_config):
```
Comment (yiming0416) [inline torchtitan/experiments/compiler_toolkit/llama3/parallelize.py:112]: we are disabling compile here anyways

File: torchtitan/experiments/compiler_toolkit/README.md (hunk 5 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/compiler_toolkit/README.md]: ok
